### PR TITLE
Add Management Server Configuration

### PR DESCRIPTION
**Pull request by StrikingThirteen**
## Overview

Description:
Add Management Server Configuration

Why should this be merged: 
Avoid directly dealing with opts.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 1]
```diff
@@ -95,8 +92,10 @@
      */
     public ManagementServer(ServerContext serverContext) {
 
-        this.opts = serverContext.getServerConfig();
-        this.localEndpoint = this.opts.get("--address") + ":" + this.opts.get("<port>");
+        this.managementServerConfig = ManagementServerConfig.parse(serverContext.getServerConfig());
+
+        this.localEndpoint = managementServerConfig.getAddress() + ":" +
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: It's much better to use NodeLocator instead of host+port, would be great if you replace host+port to NodeLocator here and in ManagementServerConfig

Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: At this point I don't want to change the fields of ManagementServer so I'll keep the endpoint as is. The ManagementServerConfiguration however is changed to be based on NodeLocator.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 2]
```diff
@@ -95,8 +93,9 @@
      */
     public ManagementServer(ServerContext serverContext) {
 
-        this.opts = serverContext.getServerConfig();
-        this.localEndpoint = this.opts.get("--address") + ":" + this.opts.get("<port>");
+        this.managementServerConfig = ManagementServerConfig.parse(serverContext.getServerConfig());
+
+        this.localEndpoint = NodeLocator.getLegacyEndpoint(managementServerConfig.getNodeLocator());
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Looks like localEndpoint is not used at all, if I am right please get rid of it

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 3]
```diff
@@ -325,4 +324,22 @@ public void shutdown() {
...
+    @Getter
+    public static class ManagementServerConfig {
+        private final NodeLocator nodeLocator;
+
+        public static ManagementServerConfig parse(Map<String, Object> opts) {
+            final String address = (String) opts.get("--address");
+            final Integer port = (Integer) opts.get("<port>");
+
+            return ManagementServerConfig.builder()
+                    .nodeLocator(NodeLocator.parseString(address + ":" + port))
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Please check if this entire class actually used by ManagementServer, looks like localEndpoint variable not used by any one.

If it's not used then just delete this class and localEndpoint parameter.
It it's used by some one then you could use `NodeLocator.builder().host().port().build()` instead of parseString

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 4]
```diff
@@ -113,15 +114,16 @@ public ExecutorService getExecutor() {
      */
     public ManagementServer(ServerContext serverContext) {
 
-        this.opts = serverContext.getServerConfig();
-        this.localEndpoint = this.opts.get("--address") + ":" + this.opts.get("<port>");
+        managementServerConfig = ManagementServerConfig.parse(serverContext.getServerConfig());
```
Comment (vjeko) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Please keep the style consistent. It is generally a good practice to prefix member variables in a constructor with "this".

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 5]
```diff
@@ -113,15 +114,16 @@ public ExecutorService getExecutor() {
...
+        localEndpoint = managementServerConfig.getAddress() + ":" +
+                             managementServerConfig.getPort();
         this.serverContext = serverContext;
 
         executor = Executors.newFixedThreadPool(serverContext.getManagementServerThreadCount(),
                 new ServerThreadFactory("management-", new ServerThreadFactory.ExceptionHandler()));
 
-        this.failureHandlerPolicy = serverContext.getFailureHandlerPolicy();
-        this.healingPolicy = serverContext.getHealingHandlerPolicy();
+        failureHandlerPolicy = serverContext.getFailureHandlerPolicy();
```
Comment (vjeko) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Why was this changed?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 6]
```diff
@@ -113,15 +114,16 @@ public ExecutorService getExecutor() {
      */
     public ManagementServer(ServerContext serverContext) {
 
-        this.opts = serverContext.getServerConfig();
-        this.localEndpoint = this.opts.get("--address") + ":" + this.opts.get("<port>");
+        managementServerConfig = ManagementServerConfig.parse(serverContext.getServerConfig());
+
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Server context already has a method to getLocalEndpoint. You can use that directly.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 2 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 3 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: moved all to sever context.
done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 4 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: I'm using "this" just when the variable in local scope shadows the fields. It seems using this adds some clutter without much value and so limiting it usage to when scope mandates its usage. Let me know if you feel strongly and will add them back.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 5 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: healingPolicy wasn't being used.
if you're referring to "this", I removed unnecessary keyword as per previous comments. similarly, if you feel strongly about the inclusion of this keyword, I'll bring them that. That being said there are other constructors in the code base where "this" only used for dealing with shadowed variables. let me know.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 6 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: yes, I think it is much cleaner to move the responsibility to ServerContext. Moved all of them there.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 5 above)
Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: I join to Vjeco and also vote for using `this` in a constructor,  just make it consistent and lined up the code. For any other methods don't use `this`

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 4 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 5 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 6 above)
Comment (StrikingThirteen) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: done

File: test/src/test/java/org/corfudb/infrastructure/ReconfigurationEventHandlerTest.java [hunk 7]
```diff
@@ -58,7 +58,7 @@ public void updateLayoutOnFailure() {
         failedServers.add(getEndpoint(SERVERS.PORT_2));
```
Comment (zalokhan) [inline test/src/test/java/org/corfudb/infrastructure/ReconfigurationEventHandlerTest.java]: Would it be possible to remove SERVERS.PORT_0 instead of 2.
This way we can test 2 more items in this test with the given policy:
1. Sequencer rollover.
2. LogUnit removal.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 8]
```diff
@@ -139,7 +142,6 @@ public ServerContext(Map<String, Object> serverConfig) {
         this.dataStore = new DataStore(serverConfig, this::dataStoreFileCleanup);
         generateNodeId();
         this.failureHandlerPolicy = new ConservativeFailureHandlerPolicy();
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: Do we want to remove this from here and move it to the ManagementServer as we discussed?
